lib/pbench/test/unit/server/test_schema.py

```diff
@@ -0,0 +1,151 @@
+import dateutil
+import pytest
+from typing import Callable
+
+from pbench.server.api.auth import Auth, UnknownUser
+from pbench.server.api.resources.query_apis import (
+    ParamType,
+    ConversionError,
+    InvalidRequestPayload,
+    MissingParameters,
+    Parameter,
+    Schema,
+)
+
+
+class TestParamType:
+    """
+    Tests on the ParamType enum
+    """
+
+    def test_enum(self):
+        assert len(ParamType.__members__) == 4
+        for n, t in ParamType.__members__.items():
+            assert str(t) == t.friendly.upper()
+            assert isinstance(t.convert, Callable)
+
+    @pytest.mark.parametrize(
+        "test",
+        (
+            (ParamType.STRING, "x", "x"),
+            (ParamType.JSON, {"key": "value"}, {"key": "value"}),
+            (ParamType.DATE, "2021-06-29", dateutil.parser.parse("2021-06-29")),
+            (ParamType.USER, "drb", "drb"),
+        ),
+    )
+    def test_successful_conversions(self, test, monkeypatch):
+        def ok(user: str) -> str:
+            return user
+
+        monkeypatch.setattr(Auth, "validate_user", ok)
+
+        ptype, value, expected = test
+        result = ptype.convert(value)
+        assert result == expected
+
+    @pytest.mark.parametrize(
+        "test",
+        (
+            (ParamType.STRING, {"not": "string"}),
+            (ParamType.JSON, "not_json"),
+            (ParamType.DATE, "2021-06-45"),
+            (ParamType.USER, "drb"),
+        ),
+    )
+    def test_failed_conversions(self, test, monkeypatch):
+        def not_ok(user: str) -> str:
+            raise UnknownUser()
+
+        monkeypatch.setattr(Auth, "validate_user", not_ok)
+
+        ptype, value = test
+        with pytest.raises(ConversionError) as exc:
+            ptype.convert(value)
+        assert str(exc).find(str(value))
+
+
+class TestParameter:
+    """
+    Tests on the Parameter class
+    """
+
+    def test_constructor(self):
+        x = Parameter("test", ParamType.STRING)
+        assert not x.required
+        assert x.name == "test"
+        assert x.type is ParamType.STRING
+
+        y = Parameter("foo", ParamType.JSON, required=True)
+        assert y.required
+        assert y.name == "foo"
+        assert y.type is ParamType.JSON
+
+    @pytest.mark.parametrize(
+        "test",
+        (
+            ({"data": "yes"}, False),
+            ({"data": None}, True),
+            ({"foo": "yes"}, True),
+            ({"foo": None, "data": "yes"}, False),
+        ),
+    )
+    def test_invalid_required(self, test):
+        x = Parameter("data", ParamType.STRING, required=True)
+        json, expected = test
+        assert x.invalid(json) is expected
+
+    @pytest.mark.parametrize(
+        "test",
+        (
+            ({"data": "yes"}, False),
+            ({"data": None}, True),
+            ({"foo": "yes"}, False),
+            ({"foo": None}, False),
```

Comment:
Orthogonal to this test, are we (meaning you ;-) ) sure that the object in line 103 should not be considered invalid?

IIRC, Elasticsearch does not treat `None` values nicely, so I wonder if we shouldn't kick back any object which contains "nulls", even if they are not in required fields...  (ES doesn't like NaN's, either....)

Reply:
If `foo` was part of the schema, (required or not) we would be complaining. however in this case we're testing essentially against a `Schema` with a single `Parameter` named `data`. My intent is that we completely ignore unexpected inputs, which allows for flexibility and API level compatibility. I don't consider failure to be a reasonable implementation of the "ignore" algorithm.

Of course we can argue this particular idea of API level compatibility; while it's served me well on previous projects, it's not without its own concerns and certainly is not in itself a panacea. In the context of this PR, however, that's how I chose to design it and this test verifies that it's behaving as intended.

Reply:
> I don't consider failure to be a reasonable implementation of the "ignore" algorithm [...] and this test verifies that it's behaving as intended.

I support you in those assertions.  However, since we're ultimately targeting Elasticsearch with this data, it might be good to have a check for nulls (and NaN's) somewhere in the pipeline; otherwise, we're going to find ourselves dealing with ingestion errors (indigestion?) at the end.  And, this seemed like a good place.

However, there is value in purity and orthogonality, so perhaps the check can be inserted somewhere else, and we can defer that until we find that it's needed.